docs(blog): introducing PWA (#53)
dhis2/developer-portal: blog/2021-11-04-introducing-pwa.md

---

> slug: introducing-pwa

Reviewer: ```suggestion
slug: 2021/11/introducing-pwa
```

title: Introducing PWA in the App Platform
author: Kai Vandivier
author_url: https://github.com/KaiVandivier

> author_image_url: https://avatars.githubusercontent.com/u/49666798?v=4

Reviewer: ```suggestion
author_image_url: https://github.com/KaiVandivier.png
```

tags: [app platform, pwa, announcement]
---

> The DHIS2 App Platform now supports PWA capabilities in apps made with the platform! The Dashboards App will be the first core app to take advantage of these features to enable offline capability, and it will be used as an example in this article to describe the details of these features.

Reviewer: I noticed that "Dashboards App" and "Dashboard App" (L108) are both used in this article, and "Dashboard" (singular) is preferred:

- https://github.com/dhis2/dashboard-app
- https://github.com/dhis2/dashboard-app/blob/master/package.json#L2
- https://github.com/dhis2/dashboard-app/blob/master/package.json#L4
- https://github.com/dhis2/dashboard-app/blob/master/d2.config.js#L4

Historically we did have a "Dashboards" app, but that was only because the name "Dashboard" was already in use. This was solved a while ago and now we only have the "Dashboard" app.

Author: Gotcha, thanks!

1. **Installability**, which means the app can be downloaded to a device and run like a native app, and
2. **Offline capability**, meaning the app can support most or all of its features while the device is offline. This works when the app is opened in a browser or as an installed app described above.

### What the Platform Provides

Apps can now be configured to take advantage of these features on an opt-in basis. Installability and offline capability, respectively, are provided by:

1. A **manifest.json** file that describes the static assets the app needs to be installed, and
2. A **service worker** script that enables offline capability by handling the app’s network traffic and caching data to serve it offline.

What’s more, we’ve developed an API in the App Runtime library that can be used in combination with these PWA features to enable **on-demand caching of individual sections of an app**. This feature is called **cacheable sections** and was developed with the Dashboards App in mind, which will use it to cache individual dashboards for offline use.

## How to use PWA in your app

> To use the basic PWA features (offline caching and installability) in an app, simply opt-in to PWA using and option in [`d2.config.js`](https://platform.dhis2.nu/#/config/d2-config-js-reference):

Reviewer 2: ```suggestion
To use the basic PWA features (offline caching and installability) in an app, simply opt-in to PWA using an option in [`d2.config.js`](https://platform.dhis2.nu/#/config/d2-config-js-reference):
```

If possible, it’s best to encourage users to save their data and update the app as soon as possible to make sure they’re using the latest version.

### Cacheable Sections

This is a set of features designed with the Dashboards app in mind to enable saving individual dashboards in the offline cache while leaving other dashboards uncached. They are designed in a generalized way to save any chunks of content with a React API, so you can use them if they suit your app too.

![Dashboards app](./assets/introducing-pwa/dashboard-save-offline.png)

The Dashboards app available in DHIS2 version 2.37 (viewable at https://play.dhis2.org/2.37dev) can be used an example of how cacheable sections work: when you visit the app and confirm any update prompts, you can go offline and still view the app. If you try to visit another dashboard though, you’ll see a “not available offline” notice.

![Dashboard not available offline](./assets/introducing-pwa/dashboard-not-available-offline.png)

> The dashboard here is the **cacheable section** – it’s only made available offline when you do so explicitly. The rest of the app that continues to work while offline is called the **app shell**, and includes content like the scripts that run the app’s behavior, some user and server data, and the user’s list of dashboards.

Reviewer 3: minor: we already use the term **app shell** to refer to the common wrapper provided by the platform.  Using this terminology again to refer to something a bit different might be confusing...

Reviewer: I was thinking that we can avoid giving it a name and gloss over it a bit:

```suggestion
The dashboard here is the **cacheable section** – it’s only made available offline when you do so explicitly. The rest of the app that continues to work while offline includes content like the scripts that run the app’s behavior, some user and server data, and the user’s list of dashboards.
```

Author: Yeah it is regrettable to have that terminology conflict; I've been using "app shell" in the caching context because it seems like that's how people refer to "the static assets and data needed to run the core of the app" in PWA contexts. If you think we need a new name for it, we could work on a better one, but will have to rename a few [config options](https://github.com/dhis2/app-platform/blob/1d0423e135b71d2005198287075e47d939040049/cli/config/d2.pwa.config.js#L25) and change some documentation

What do you think about adding some more description to make it clearer that "app shell" here is referring to the cache contents, not to be confused with the common wrapper? Or removing the reference to "app shell" here like Viktor suggests?

To learn more about the usage of cacheable sections, including an example of a Loading Mask, check out the [“Usage” section](https://runtime.dhis2.nu/#/advanced/offline?id=usage) in the App Runtime docs.

### Online status

If you took a look at the 2.37 version of the Dashboard app while offline, you may have noticed that many features of the app are disabled while offline, and some items in the interface change appearance.

The App Runtime provides a `useOnlineStatus` hook that can access the online or offline status of the app, which the Dashboard app uses to enable or disable those features. Using the hook looks like this:

```jsx
import { useOnlineStatus } from "@dhis2/app-runtime";

const OnlineComponent = () => {
  const { online, offline } = useOfflineStatus();

  return <p>{online ? "Online" : "Offline"}</p>;
};
```

## An important caveat

These features are exciting and the opportunity to take apps offline can be very useful when taken into areas with low network connectivity, but there is an important limitation that you should know before using the PWA features.

> In its current implementation, **cached data is not encrypted** when stored offline, and a malicious actor could inspect access and inspect the cached data if they gain access to a user’s device. For this reason, it’s important to **not use** PWA features in an app that will handle sensitive data, especially if the app is likely to be used on shared devices.

Reviewer 3: ```suggestion
In its current implementation, **cached data is not encrypted** when stored offline, and a malicious actor could inspect access and inspect the cached data if they gain access to a user’s device. For this reason, it is **not recommended** to use PWA features in an app that will handle sensitive data, especially if the app is likely to be used on shared devices.  Some protections are currently provided by the platform to guard against unauthorized access to data when a user logs out or when a new user logs in, but without encryption it might still be possible for a savvy attacker to access some offline data.  Importantly, the data cached by the PWA is the same as might be cached at the browser layer in a non-PWA application, and so the risk of data exposure has not significantly increased.
```

Encryption of the offline data will be the next feature for these PWA tools however, so keep an eye out for the announcement of that release.

## Want to know more?

If you would like to learn more about adding offline caching, cacheable sections, or using other PWA features to your app, take a look at the documentation at the [App Platform](https://platform.dhis2.nu/#/pwa/pwa) and [App Runtime](https://runtime.dhis2.nu/#/advanced/offline) sites which go into greater detail about the API and provide more examples.

> In the near future we will also post an detailed article about the technical development of these PWA features because we’re using these tools in a pretty unique and cool way that we want to share, so keep an eye out for that post!

Reviewer 3: ```suggestion
In the near future we will also post a detailed article about the technical development of these PWA features because we’re using these tools in a pretty unique and cool way that we want to share, so keep an eye out for that post!
```
